Fix part #549 :Options low fi Navigation Drawer App module

veena14cs opened the pull request:
## Explanation
This PR focuses on the app module work for user preference screen to manage the behaviour of the App.

Steps to follow
- Create new profile to see default options.

### Mock
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/2776cbb1-22b0-42c9-b70b-2136bdb361da/OP-Options-Menu-/

### Design Doc
https://docs.google.com/document/d/1uylgD2SwxHMYmHgof_-OTlUwriBL5Ps17r0hEWI0icg/edit#

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

veena14cs committed:
Added option screen

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
working on bindable adapter

veena14cs committed:
Update OptionsFragment.kt

veena14cs committed:
adde test cases

veena14cs committed:
Done nit changes

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
renamed toolbar

veena14cs committed:
Update OptionsFragmentTest.kt

veena14cs committed:
Update OptionsFragmentTest.kt

veena14cs committed:
nit change

veena14cs committed:
fixed nit

veena14cs committed:
working on shared pref

veena14cs committed:
Update OptionsFragment.kt

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
working on domain layer

veena14cs committed:
fixed livedata issue.

veena14cs committed:
Update OptionsFragment.kt

veena14cs committed:
Update AndroidManifest.xml

veena14cs committed:
Update OptionsFragment.kt

veena14cs committed:
Fixed nit

veena14cs committed:
added test cases

veena14cs committed:
Update OptionsFragment.kt

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
Update ActivityComponent.kt

veena14cs committed:
added options in navigation

veena14cs committed:
added icon for options

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
added currentprofileId

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
removed app module from this pr

veena14cs committed:
removed layouts

veena14cs committed:
Update misc.xml

veena14cs committed:
updated parameters

veena14cs committed:
Update ProfileManagementControllerTest.kt

veena14cs committed:
Update ProfileManagementControllerTest.kt

veena14cs committed:
update

veena14cs committed:
Update navigation_drawer_menu.xml

veena14cs committed:
Merge branch 'options-low-fi' into options-low-fi-part2

veena14cs committed:
updated option preferences

veena14cs committed:
updated layouts

veena14cs committed:
add spacing

veena14cs committed:
fixed spacing

veena14cs committed:
add spacing

veena14cs committed:
add default values

veena14cs committed:
Update ProfileActivityPresenter.kt

veena14cs committed:
Merge branch 'options-low-fi' into options-low-fi-part2

veena14cs committed:
Update ic_options_icon_grey_24dp.xml

veena14cs committed:
Update strings.xml

veena14cs committed:
Create OptionsFragmentTest.kt

veena14cs committed:
Update ProfileManagementController.kt

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
Merge branch 'options-low-fi' into options-low-fi-part2

veena14cs committed:
fixing profileid

veena14cs committed:
Update OptionsFragmentTest.kt

rt4914 commented on app/build.gradle:
```diff
@@ -72,6 +72,7 @@ dependencies {
       "androidx.fragment:fragment:$fragment_version",
       'androidx.lifecycle:lifecycle-runtime-ktx:2.2.0-alpha03',
       'androidx.multidex:multidex:2.0.1',
+      'androidx.preference:preference:1.1.0-alpha02',
```
As per this link: https://developer.android.com/jetpack/androidx/releases/preference

The latest version is `1.1.0`, so please update this.

veena14cs replied:
Updated.

rt4914 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -29,6 +29,22 @@
+      android:name=".options.OptionsActivity"
+      android:screenOrientation="portrait" />
+    <activity
+
```
Remove this blank line

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/drawer/NavigationDrawerFragmentPresenter.kt:
```diff
@@ -14,23 +15,31 @@ import org.oppia.app.R
   private lateinit var drawerLayout: DrawerLayout
   private var previousMenuItemId: Int? = null
+  private var internalProfileId: Int = -1
+  private lateinit var profileId: ProfileId
```
This is unused and not needed.

veena14cs replied:
Its getting used.

rt4914 commented on app/src/main/java/org/oppia/app/drawer/NavigationDrawerFragmentPresenter.kt:
```diff
@@ -14,23 +15,31 @@ import org.oppia.app.R
     navView.setNavigationItemSelectedListener(this)
     fragment.setHasOptionsMenu(true)
+    internalProfileId = activity.intent.getIntExtra(KEY_HOME_PROFILE_ID, -1)
+    profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
```
This is unused and not needed.

rt4914 replied:
If you remove this, remove the unused import too.

rt4914 commented on app/src/main/java/org/oppia/app/drawer/NavigationDrawerFragmentPresenter.kt:
```diff
@@ -40,6 +49,9 @@ class NavigationDrawerFragmentPresenter @Inject constructor(
           Intent(fragment.activity, HomeActivity::class.java)
         }
+        NavigationDrawerItem.OPTIONS -> {
+          OptionsActivity.createOptionsActivity(fragment.requireContext(),internalProfileId)
```
Spacing issue `... requireContext(), internalProfileId)`

veena14cs replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/app/options/LanguageSelectionAdapter.kt:
```diff
@@ -0,0 +1,67 @@
+          val intent = Intent()
+          intent.putExtra("MESSAGE", language)
+          (binding.radioContainer.context as AppLanguageActivity).setResult(2, intent)
+          (binding.radioContainer.context as AppLanguageActivity).finish()//finishing activity
```
Remove comment `//finishing activity`

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/LanguageSelectionAdapter.kt:
```diff
@@ -0,0 +1,67 @@
+          val intent = Intent()
+          intent.putExtra("MESSAGE", language)
+          (binding.radioContainer.context as DefaultAudioActivity).setResult(3, intent)
+          (binding.radioContainer.context as DefaultAudioActivity).finish()//finishing activity
```
Ditto

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsActivityPresenter.kt:
```diff
@@ -0,0 +1,30 @@
+class OptionsActivityPresenter @Inject constructor(
+  private val activity: AppCompatActivity,
+  private val logger: Logger,
+  private val profileManagementController: ProfileManagementController) {
```
Update:
```
  private val profileManagementController: ProfileManagementController
) {
```

rt4914 replied:
The brackets should come to next line

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsActivityPresenter.kt:
```diff
@@ -0,0 +1,30 @@
+    if (getOptionFragment() == null) {
+      activity.supportFragmentManager.beginTransaction().add(
+        R.id.option_fragment_placeholder,
+        OptionsFragment(activity,profileManagementController,logger)
```
Spacing issues: ` OptionsFragment(activity, profileManagementController, logger)`

rt4914 replied:
Getting this warning in add function:

<img width="826" alt="Screenshot 2020-02-03 at 11 18 32 AM" src="https://user-images.githubusercontent.com/9396084/73628681-f8a6fd00-4676-11ea-9f41-aefea392d051.png">

veena14cs replied:
I checked this thing. I am not able to see any warning. Can you please check it once again.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,185 @@
+  private lateinit var profileId: ProfileId
+  var storyTextSize = 16f
+  var appLanguage = "English"
+  var audioLanguage = "No Audio"
```
Make all these 3 `var`'s private

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,185 @@
+          preference.summary = stringValue
+          profileManagementController.updateAppLanguage(profileId, stringValue)
+        }
+        preference.key == getString(R.string.key_default_audio) -> // update the changed audio language  to summary filed
```
Prefer writing comments above this line. as such comments can easily get skipped.

rt4914 replied:
Also, write proper comment meaning the spacing between words should be single space, first character should be capital, full-stop to end the sentence.

veena14cs replied:
Corrected comment line.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,185 @@
+    }
+
+  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+    // Check which request we're responding to
```
Remove this comment, not-needed as it is self explanatory

veena14cs replied:
Removed.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,185 @@
+    // Check which request we're responding to
+    when (requestCode) {
+      1 -> {
+        val textSize = data!!.getStringExtra("MESSAGE") as String
```
Why are these keys same, they should be different with proper context like `KEY_TEXT_SIZE`, also, create then as constants and use those keys wherever needed.

veena14cs replied:
Created constants to pass through intents.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,185 @@
+      storyTextSize = it.storyTextSize
+      appLanguage = it.appLanguage
+      audioLanguage = it.audioLanguage
+      logger.e("OptionsFragment", "Failed to retrieve profile"+ storyTextSize)
```
Update: `logger.e("OptionsFragment", "Failed to retrieve profile$storyTextSize")`

veena14cs replied:
Removed this log as it was added for debugging.

rt4914 commented on app/src/main/java/org/oppia/app/options/StoryTextSizeActivityPresenter.kt:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    when (prefSummaryValue) {
+      "Small" -> {
```
Can we define these values somewhere else in such a way that only constants are used and the hardcoded strings can be avoided in such critical places.

veena14cs replied:
Made use of constants.

rt4914 reviewed (changes requested):
I have taken a brief look at this PR.

Please add toolbar to OptionsActivity which is linked to navigation drawer and has correct title too [check mocks for more details: https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/2776cbb1-22b0-42c9-b70b-2136bdb361da/OP-Options-Menu-/].

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
Merge branch 'options-low-fi' into options-low-fi-part2

veena14cs committed:
Create NavigationDrawerHeaderViewModel.kt

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
Merge branch 'options-low-fi' into options-low-fi-part2

veena14cs committed:
fixing issues.

veena14cs committed:
Fixed issues.

veena14cs committed:
Update OptionsFragment.kt

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/AppLanguageActivityPresenter.kt:
```diff
@@ -0,0 +1,46 @@
+    languageList.add("Hindi")
+    languageList.add("Chinese")
+    languageSelectionAdapter.setLanguageList(languageList)
+
```
remove empty line

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/DefaultAudioActivityPresenter.kt:
```diff
@@ -0,0 +1,48 @@
+    languageList.add("Hindi")
+    languageList.add("Chinese")
+    languageSelectionAdapter.setLanguageList(languageList)
+
```
ditto.

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/main/res/layout/story_text_size_activity.xml:
```diff
@@ -0,0 +1,119 @@
+      </LinearLayout>
+
+      <View
+        android:id="@+id/toolbar_shadow_view"
```
rename as per file

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/main/res/layout/story_text_size_activity.xml:
```diff
@@ -0,0 +1,119 @@
+          android:text="@string/story_text_will_look_like_this" />
+
+        <View
+          android:id="@+id/divider_view"
```
If its used rename as per file name else please remove.

veena14cs replied:
done.

nikitamarysolomanpvt commented on app/src/main/res/layout/app_language_activity.xml:
```diff
@@ -0,0 +1,62 @@
+        app:layout_constraintTop_toBottomOf="@id/toolbar_shadow_view" />
+
+      <View
+        android:id="@+id/toolbar_shadow_view"
```
ditto

veena14cs replied:
done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,197 @@
+const val MESSAGE_APP_LANGUAGE = "App Language"
+const val MESSAGE_AUDIO_LANGUAGE = "Audio Language"
+
+class OptionsFragment @Inject constructor(
```
`Options + Cmd + L` to fix the formatting issues in this file.

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/StoryTextSizeActivity.kt:
```diff
@@ -0,0 +1,32 @@
+import org.oppia.app.activity.InjectableAppCompatActivity
+import javax.inject.Inject
+
+private const val STORY_TEXT_SIZE_PREFERENCE_SUMMARY_VALUE = "STORY_TEXT_SIZE_PREFERENCE_SUMMARY_VALUE"
```
Please rename the key name as STORY_TEXT_SIZE_PREFERENCE_SUMMARY_KEY

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/StoryTextSizeActivityPresenter.kt:
```diff
@@ -0,0 +1,87 @@
+    when (prefSummaryValue) {
+      STORY_TEXT_SIZE_SMALL -> {
+        binding.storyTextSizeSeekBar.progress = 0
+        binding.previewTextview.textSize = 16f
```
The constant values in this file are getting reused and if by mistake we change 16f to 15f in one place and forget about the second place it will bring inconsistency. 

So define these constants at the top like:

`private const val SMALL_TEXT_SIZE_VALUE = 16f`

veena14cs replied:
Done thanks.

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/AppLanguageActivity.kt:
```diff
@@ -0,0 +1,39 @@
+import org.oppia.app.activity.InjectableAppCompatActivity
+import javax.inject.Inject
+
+private const val APP_LANGUAGE_PREFERENCE_KEY = "APP_LANGUAGE_PREFERENCE_KEY"
```
Please remove KEY from string as "APP_LANGUAGE_PREFERENCE"

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/AppLanguageActivity.kt:
```diff
@@ -0,0 +1,39 @@
+import javax.inject.Inject
+
+private const val APP_LANGUAGE_PREFERENCE_KEY = "APP_LANGUAGE_PREFERENCE_KEY"
+private const val APP_LANGUAGE_PREFERENCE_SUMMARY_VALUE = "APP_LANGUAGE_PREFERENCE_SUMMARY_VALUE"
```
ditto

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
I think you missed by top level comment in this PR.

https://github.com/oppia/oppia-android/pull/614#pullrequestreview-352028112

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/DefaultAudioActivity.kt:
```diff
@@ -0,0 +1,36 @@
+import org.oppia.app.activity.InjectableAppCompatActivity
+import javax.inject.Inject
+
+private const val AUDIO_LANGUAGE_PREFERENCE_KEY = "AUDIO_LANGUAGE_PREFERENCE_KEY"
```
ditto

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/DefaultAudioActivity.kt:
```diff
@@ -0,0 +1,36 @@
+import javax.inject.Inject
+
+private const val AUDIO_LANGUAGE_PREFERENCE_KEY = "AUDIO_LANGUAGE_PREFERENCE_KEY"
+private const val AUDIO_LANGUAGE_PREFERENCE_SUMMARY_VALUE = "AUDIO_LANGUAGE_PREFERENCE_SUMMARY_VALUE"
```
ditto

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/AppLanguageActivityPresenter.kt:
```diff
@@ -0,0 +1,46 @@
+    val languageList = ArrayList<String>()
+    languageList.add("English")
+    languageList.add("French")
+    languageList.add("Hindi")
```
Suggestion language name should be in respective languages... so that those who don't know english also can read.

veena14cs replied:
This is the dummy list. Not sure whether this list will be coming from backend. I will add a TODO comment.

rt4914 replied:
Maybe confirm about this with Ben.

nikitamarysolomanpvt replied:
> This is the dummy list. Not sure whether this list will be coming from backend. I will add a TODO comment.

Okay please do confirm with ben

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/LanguageSelectionAdapter.kt:
```diff
@@ -0,0 +1,67 @@
+          (binding.radioContainer.context as AppLanguageActivity).finish()
+        } else {
+          val intent = Intent()
+          intent.putExtra(MESSAGE_AUDIO_LANGUAGE, language)
```
Rename as KEY

veena14cs replied:
Added prefix key to const variable.

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/LanguageSelectionAdapter.kt:
```diff
@@ -0,0 +1,67 @@
+import org.oppia.app.databinding.LanguageItemsBinding
+
+// TODO(#216): Make use of generic data-binding-enabled RecyclerView adapter.
+/** Adapter to bind languages to [RecyclerView] inside [AppLanguageActivityPresenter] and [DefaultAudioActivityPresenter]. */
```
Please leave an empty line after TODO

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/options/OptionsActivityPresenter.kt:
```diff
@@ -0,0 +1,31 @@
+  fun handleOnCreate() {
+    activity.setContentView(R.layout.option_activity)
+    if (getOptionFragment() == null) {
+      activity.supportFragmentManager.beginTransaction().add(
```
Getting error as mentioned by Rajat

veena14cs replied:
This seems to be gradle build issue. If you clean and rebuild, warning is not shown. Tried it in Nikita's system.

veena14cs committed:
setupnavigation drawer

veena14cs commented:
> I think you missed by top level comment in this PR.
> 
> [#614 (review)](https://github.com/oppia/oppia-android/pull/614#pullrequestreview-352028112)

Yes I missed it. Added now PTAL.

veena14cs committed:
nit

rt4914 commented:
> > I think you missed by top level comment in this PR.
> > [#614 (review)](https://github.com/oppia/oppia-android/pull/614#pullrequestreview-352028112)
> 
> Yes I missed it. Added now PTAL.

@veena14cs Please add test case for click on navigation drawer too.

nikitamarysolomanpvt reviewed (approved):
Some nit changes PTAL. Beside that PR LGTM

veena14cs committed:
Fixed nit issues

veena14cs committed:
nit change

veena14cs committed:
Update OptionsFragmentTest.kt

veena14cs committed:
Update OptionsFragmentTest.kt

veena14cs commented:
> > > I think you missed by top level comment in this PR.
> > > [#614 (review)](https://github.com/oppia/oppia-android/pull/614#pullrequestreview-352028112)
> > 
> > 
> > Yes I missed it. Added now PTAL.
> 
> @veena14cs Please add test case for click on navigation drawer too.

Added test case for click on navigation drawer. PTAL

rt4914 commented on app/src/sharedTest/java/org/oppia/app/options/OptionsFragmentTest.kt:
```diff
@@ -0,0 +1,300 @@
+  @Test
+  fun testOptionFragment_clickNavigationDrawerHamburger_navigationDrawerIsOpenedSuccessfully() {
+    launch<OptionsActivity>(createOptionActivityIntent(0)).use {
+    onView(ViewMatchers.withContentDescription(R.string.drawer_open_content_description)).check(
```
Everywhere in this file import ViewMatchers.withContentDescription and similar method directly, for better code clarity.

Sample:

```
  @Test
  fun testOptionFragment_clickNavigationDrawerHamburger_navigationDrawerIsOpenedSuccessfully() {
    launch<OptionsActivity>(createOptionActivityIntent(0)).use {
      onView(withContentDescription(R.string.drawer_open_content_description)).check(matches(isCompletelyDisplayed()))
        .perform(click())
      onView(withId(R.id.options_fragment_placeholder)).check(matches(isCompletelyDisplayed()))
      onView(withId(R.id.options_activity_drawer_layout)).check(matches(isOpen()))
    }
  }
```

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
Everything looks good but there is still one issue:

1. Open OptionsActivity.
2. Click back button -> It opens ProfileSelectionScreen but instead it should go to HomeScreen.

veena14cs committed:
fixed navigation

veena14cs commented:
> Everything looks god but there is still one issue:
> 
> 1. Open OptionsActivity.
> 2. Click back button -> It opens ProfileSelectionScreen but instead it should go to HomeScreen.

Fixed navigation please check.

rt4914 reviewed (approved):
LGTM, thanks

rt4914 commented:
@veena14cs Do not merge this, it is branched from `options-low-fi`.

veena14cs commented:
> @veena14cs Do not merge this, it is branched from `options-low-fi`.

Ack.

veena14cs committed:
Update AddProfileActivityPresenter.kt

veena14cs committed:
update proto

veena14cs committed:
Merge branch 'develop' into options-low-fi

veena14cs committed:
added func to get storytext size

veena14cs committed:
Merge branch 'options-low-fi' into options-low-fi-part2

veena14cs committed:
updated app module

veena14cs committed:
updated text size.

rt4914 commented:
@veena14cs Unassigning Ben' from this one considering that the previous branch is not yet in develop.

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

veena14cs committed:
added domain layer code

veena14cs committed:
nit change

veena14cs committed:
fixes issues

veena14cs committed:
Update NavigationDrawerFragmentPresenter.kt

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

rt4914 commented:
@veena14cs Please check non-flaky tests first, they are failing.

veena14cs commented:
> @veena14cs Unassigning Ben' from this one considering that the previous branch is not yet in develop.

Domain layer is been merged to develop. Please check once again.

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

veena14cs committed:
resolved conflicts

rt4914 commented on app/src/main/java/org/oppia/app/drawer/NavigationDrawerFragmentPresenter.kt:
```diff
@@ -97,16 +98,21 @@ class NavigationDrawerFragmentPresenter @Inject constructor(
     if (previousMenuItemId != menuItemId && menuItemId != 0) {
       when (NavigationDrawerItem.valueFromNavId(menuItemId)) {
         NavigationDrawerItem.HOME -> {
+          drawerLayout.closeDrawers()
```
For now, please don't bring any change to other options in nav drawer, just focus on "OPTIONS" only, because they are not working correctly.

rt4914 commented on app/src/main/java/org/oppia/app/drawer/NavigationDrawerFragmentPresenter.kt:
```diff
@@ -97,16 +98,21 @@ class NavigationDrawerFragmentPresenter @Inject constructor(
+        }
+        NavigationDrawerItem.OPTIONS -> {
+          drawerLayout.closeDrawers()
+          fragment.activity!!.startActivity(OptionsActivity.createOptionsActivity(fragment.requireContext(), internalProfileId))
```
Open Options from nav-drawer, click back -> home appears and now open drawer again, you will see `Options` highlighted. Instead, home should be highlighted. Once you fix this, also write test case to make sure this works correctly.

nikitamarysolomanpvt replied:
+1

veena14cs replied:
I will remove those changes. When we press back from Option Page it is directly going to Profile page instead of HomePage to fix it I added this code. But this is fixed and its updated in develop.

rt4914 commented on app/src/main/java/org/oppia/app/options/AppLanguageActivity.kt:
```diff
@@ -0,0 +1,47 @@
+    val message = prefSummaryValue
+    val intent = Intent()
+    intent.putExtra(KEY_MESSAGE_APP_LANGUAGE, message)
+    setResult(2, intent)
```
Add inline comment here `/* resultCode= */` to make first parameter clear. Ditto in otehr places too.

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/AppLanguageActivityPresenter.kt:
```diff
@@ -0,0 +1,46 @@
+@ActivityScope
+class AppLanguageActivityPresenter @Inject constructor(private val activity: AppCompatActivity) {
+  private lateinit var languageSelectionAdapter: LanguageSelectionAdapter
+  private var prefSummaryValue: String? = null
```
Make this lateint var

nikitamarysolomanpvt replied:
+1

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/DefaultAudioActivity.kt:
```diff
@@ -0,0 +1,44 @@
+    val message = prefSummaryValue
+    val intent = Intent()
+    intent.putExtra(KEY_MESSAGE_AUDIO_LANGUAGE, message)
+    setResult(3, intent)
```
Ditto as before

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/DefaultAudioActivityPresenter.kt:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  private fun createAdapter() {
+    // TODO: Replace dummy list with actual language list.
```
Link this TODO with an issue number

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/drawer/NavigationDrawerItem.kt:
```diff
@@ -4,7 +4,7 @@ import org.oppia.app.R
 /** Represents different destinations that can be selected by the user from the navigation drawer. */
 enum class NavigationDrawerItem(val value: Int) {
-  HOME(R.id.nav_home), HELP(R.id.nav_help), SWITCH_PROFILE(R.id.nav_switch_profile);
+  HOME(R.id.nav_home),OPTIONS(R.id.nav_options), HELP(R.id.nav_help), SWITCH_PROFILE(R.id.nav_switch_profile);
```
Nit: space before ` OPTIONS`

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,261 @@
+import org.oppia.util.logging.Logger
+import javax.inject.Inject
+
+const val KEY_MESSAGE_STORY_TEXT_SIZE = "Text Size"
```
generally we are following this convention: `text_size` or `TEXT_SIZE` but not this one.

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,261 @@
+
+  override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
+    setPreferencesFromResource(R.xml.basic_preference, rootKey)
+    internalProfileId = activity.intent.getIntExtra(KEY_NAVIGATION_PROFILE_ID, -1)
```
if the id is `-1` how will the application respond ?

veena14cs replied:
It is getting id from Profilechooser. It cannot have -1.

rt4914 replied:
ok.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,261 @@
+            requireContext(),
+            getString(R.string.key_story_text_size),
+            textSizePref!!.summary.toString()
+          ), 1
```
this `1` should be replaced with a constant value 

`private const val REQUEST_CODE_TEXT_SIZE = 1`

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,261 @@
+
+  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+    when (requestCode) {
+      1 -> {
```
Ditto. Use const here.

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/StoryTextSizeActivity.kt:
```diff
@@ -0,0 +1,47 @@
+    val message = prefSummaryValue
+    val intent = Intent()
+    intent.putExtra(KEY_MESSAGE_STORY_TEXT_SIZE, message)
+    setResult(1, intent)
```
Ditto as earlier

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsActivity.kt:
```diff
@@ -0,0 +1,29 @@
+      return intent
+    }
+  }
+  override fun onCreate(savedInstanceState: Bundle?) {
```
Add one blank line above this.

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
Some nit changes. Also, I think this PR should be assigned to Ben also, as there are some places where he might have an alternate opinion.

rt4914 reviewed (commented):
Some nit changes. Also, I think this PR should be assigned to Ben also, as there are some places where he might have an alternate opinion.

veena14cs committed:
updated nit

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

veena14cs committed:
Update NavigationDrawerFragmentPresenter.kt

rt4914 commented on app/src/main/java/org/oppia/app/options/AppLanguageActivity.kt:
```diff
@@ -0,0 +1,47 @@
+    val message = prefSummaryValue
+    val intent = Intent()
+    intent.putExtra(KEY_MESSAGE_APP_LANGUAGE, message)
+    setResult(/* resultCode= */ REQUEST_CODE_APP_LANGUAGE, intent)
```
This `/* resultCode= */ ` was needed only when there was number directly getting used. Now it is not needed, remove it.

rt4914 replied:
Ditto elsewhere.

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/LanguageSelectionAdapter.kt:
```diff
@@ -0,0 +1,68 @@
+        if (prefKey == binding.radioContainer.context.getString(R.string.key_app_language)) {
+          val intent = Intent()
+          intent.putExtra(KEY_MESSAGE_APP_LANGUAGE, language)
+          (binding.radioContainer.context as AppLanguageActivity).setResult(/* resultCode= */ 2, intent)
```
Use those constants that you created in other file here. Instead of `/* resultCode= */ 2`

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/LanguageSelectionAdapter.kt:
```diff
@@ -0,0 +1,68 @@
+        } else {
+          val intent = Intent()
+          intent.putExtra(KEY_MESSAGE_AUDIO_LANGUAGE, language)
+          (binding.radioContainer.context as DefaultAudioActivity).setResult(/* resultCode= */ 3, intent)
```
Ditto

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,264 @@
+          }
+        }
+      }
+      true
```
Add inline comment for this true. What is it?

veena14cs replied:
It is just Boolean to update state. Added inline comment.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragment.kt:
```diff
@@ -0,0 +1,264 @@
+        bindPreferenceSummaryToValue(audioLanguage, findPreference(getString(R.string.key_default_audio)))
+      }
+    }
+
```
Remove this blank line

nikitamarysolomanpvt replied:
+1

veena14cs replied:
Done.

rt4914 reviewed (approved):
Suggested some changes, also, once everything is fixed, suggestion adding Ben to review this PR.

nikitamarysolomanpvt reviewed (approved):
LGTM please address the nit changes suggestions.

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

veena14cs committed:
updated nit change.

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

veena14cs committed:
changed implementation to recyclerview

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

veena14cs committed:
updated implementation

veena14cs committed:
Update OptionsFragmentPresenter.kt

veena14cs committed:
working on live data

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

veena14cs committed:
added live data in view model

veena14cs committed:
resolving issue

veena14cs committed:
removed commented code

veena14cs committed:
updated options fragment

veena14cs commented:
As per @BenHenning suggestion on one of the thread, I have changed implementation of Options Users Preference settings from using Preferences API to Recyclerview's implementation. So I am reassigning it to @rt4914 and @nikitamarysolomanpvt. PTAL once again.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsActivityPresenter.kt:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  fun updateStoryTextSize(textSize: String) {
+     getOptionFragment()?.updateStoryTextSize(textSize)
```
Nit: spacing issue

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsAppLanguageViewModel.kt:
```diff
@@ -0,0 +1,6 @@
+
+/** App language settings view model for the recycler view in [OptionsFragment]. */
+class OptionsAppLanguageViewModel : OptionsItemViewModel() {
+  var appLanguage: String = ""
```
Instead of var make it val `ObservableField`
Ditto elsewhere.

```
/** App language settings view model for the recycler view in [OptionsFragment]. */
class OptionsAppLanguageViewModel : OptionsItemViewModel() {
  val appLanguage = ObservableField<String>("")

  fun setAppLanguage(appLanguageValue: String) {
    appLanguage.set(appLanguageValue)
  }
}

```

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsAudioLanguageViewModel.kt:
```diff
@@ -0,0 +1,7 @@
+class OptionsAudioLanguageViewModel : OptionsItemViewModel() {
+  var audioLanguage: String = ""
+}
+
```
Remove this extra blank line

veena14cs replied:
Removed.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsStoryTextViewViewModel.kt:
```diff
@@ -0,0 +1,7 @@
+class OptionsStoryTextViewViewModel : OptionsItemViewModel() {
+  var storyTextSize: String = ""
+}
+
```
Ditto

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/values/strings.xml:
```diff
@@ -223,4 +224,16 @@
   <string name="back_button">Back Button</string>
   <string name="next">Next</string>
+  <!-- OptionsActivity -->
+  <string name="key_story_text_size">key_story_text_size</string>
```
Key should not come in `strings.xml` file rather they should be const vals in kotlin files because these keys are not needed to be translated.

nikitamarysolomanpvt replied:
+1

veena14cs replied:
Done,

rt4914 commented on domain/src/main/java/org/oppia/domain/profile/ProfileManagementController.kt:
```diff
@@ -386,6 +386,7 @@ class ProfileManagementController @Inject constructor(
         it,
         ProfileActionStatus.PROFILE_NOT_FOUND
       )
+      logger.d("text size","======="+storyTextSize)
```
Remove this.

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
Suggested some changes. Also add test cases.

nikitamarysolomanpvt reviewed (commented):
Please add test cases and assign me back

veena14cs committed:
fixed issues

veena14cs committed:
fixed spacing

veena14cs committed:
Fixed testcases

rt4914 commented on app/src/main/java/org/oppia/app/options/StoryTextSizeActivity.kt:
```diff
@@ -0,0 +1,49 @@
+class StoryTextSizeActivity : InjectableAppCompatActivity() {
+
+  @Inject
+  lateinit var storyTextSizeActivityPresenter: StoryTextSizeActivityPresenter
```
Single line:
`@Inject  lateinit var storyTextSizeActivityPresenter: StoryTextSizeActivityPresenter`

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
StoryTextSize is not getting saved.

Also, the test cases are just checking selection is done correctly on UI but it does not make sure that they are getting saved. 

Better approach should be, example:
Open Audio Language Selection -> Select Language -> Go Back -> Check if the language is shown -> Open Audio Language Selection -> Check if the last selected language is shown.

nikitamarysolomanpvt reviewed (changes requested):
Text size not working PTAL

veena14cs committed:
fixed language selection

veena14cs committed:
fixing app language testcase.

veena14cs committed:
fixed testcases

veena14cs committed:
nit change

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

rt4914 reviewed (changes requested):
I have found following issues in app:
1. You select any language or textsize, now if you click on back button in toolbar, it is getting saved but if you click on back button present in default android OS, it is not working.
2. From OptionsActivity, if go back it opens `ProfileChooser` but instead `Home` should be selected.

nikitamarysolomanpvt commented:
> I have found following issues in app:
> 
> 1. You select any language or textsize, now if you click on back button in toolbar, it is getting saved but if you click on back button present in default android OS, it is not working.
> 2. From OptionsActivity, if go back it opens `ProfileChooser` but instead `Home` should be selected.

+1

veena14cs commented:
> I have found following issues in app:
> 
> 2. From OptionsActivity, if go back it opens `ProfileChooser` but instead `Home` should be selected.

This needs to be fixed in Navigation drawer. This Pr is focusing only on Options screen. Should I fix it in navigation drawer only for options?

veena14cs committed:
Fixed issues

veena14cs committed:
Update OptionsFragmentTest.kt

veena14cs commented:
> I have found following issues in app:
> 
> 1. You select any language or textsize, now if you click on back button in toolbar, it is getting saved but if you click on back button present in default android OS, it is not working.

Fixed this issue.

rt4914 commented:
> > I have found following issues in app:
> > 
> > 1. From OptionsActivity, if go back it opens `ProfileChooser` but instead `Home` should be selected.
> 
> This needs to be fixed in Navigation drawer. This Pr is focusing only on Options screen. Should I fix it in navigation drawer only for options?

I think it is a part of OptionsActivity. Anyways, if you plan to solve this in separate PR, then file and issue and link it here.

rt4914 commented on app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt:
```diff
@@ -0,0 +1,244 @@
+
+  private fun bindStoryTextSize(binding: OptionStoryTextSizeBinding, model: OptionsStoryTextViewViewModel) {
+    binding.viewModel = model
+    binding.storyTextSizeTextView.setOnClickListener {
```
Just noticed this major issue:

Actually the click listener should be set on the entire item and not just the text.

You will need to update the test cases too.

veena14cs replied:
I have updated listener and testcases related to it. PTAL.

rt4914 reviewed (changes requested):
PTAL.

veena14cs commented:
> > > I have found following issues in app:
> > > 
> > > 1. From OptionsActivity, if go back it opens `ProfileChooser` but instead `Home` should be selected.
> > 
> > 
> > This needs to be fixed in Navigation drawer. This Pr is focusing only on Options screen. Should I fix it in navigation drawer only for options?
> 
> I think it is a part of OptionsActivity. Anyways, if you plan to solve this in separate PR, then file and issue and link it here.

I have fixed it in the same PR. PTAL

veena14cs committed:
fixed item listener

veena14cs committed:
Merge branch 'develop' into options-low-fi-part2

rt4914 reviewed (approved):
LGTM, thanks.

nikitamarysolomanpvt reviewed (approved):
LGTM

veena14cs committed:
Update StoryTextSizeActivityPresenter.kt
